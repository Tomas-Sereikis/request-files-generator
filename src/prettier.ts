import * as prettier from 'prettier'
import { trim } from 'lodash'

const prettierrc = require('../.prettierrc.json')

export function formatCode(code: string): string {
  return trim(
    prettier.format(code, {
      parser: 'typescript',
      ...prettierrc,
    }),
    '\n',
  )
}
