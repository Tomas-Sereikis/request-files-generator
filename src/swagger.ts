import * as swaggerParser from 'swagger-parser'
import * as swaggerScheme from 'swagger-schema-official'
import { assign, flatten } from 'lodash'

export async function validate(docs: swaggerScheme.Spec): Promise<ISwaggerDocs> {
  return await swaggerParser.validate(docs)
}

export function swaggerSchemeObjectPropertiesMerge(object: ISwaggerSchemeObject): ISwaggerSchemeObjectProperties {
  const type = SwaggerSchemeType.object
  if (object.allOf) {
    const required = flatten(object.allOf.map(curr => curr.required))
    const properties = assign({} as ISwaggerSchemeObjectProperty, ...object.allOf.map(curr => curr.properties))
    return { properties, required, type }
  } else if (object.properties && object.required) {
    return { properties: object.properties, required: object.required, type }
  } else {
    return { properties: {}, required: [], type }
  }
}

export enum SwaggerSchemeType {
  object = 'object',
  boolean = 'boolean',
  array = 'array',
  string = 'string',
  number = 'number',
  integer = 'integer',
  interface = 'interface',
}

export enum SwaggerParameterIn {
  query = 'query',
  path = 'path',
  body = 'body',
}

export interface ISwaggerDocs {
  info: ISwaggerDocsInfo
  paths: ISwaggerPaths
}

export interface ISwaggerDocsInfo {
  title: string
  description?: string
}

export interface ISwaggerPaths {
  [url: string]: ISwaggerPath
}

export interface ISwaggerPath {
  [method: string]: ISwaggerPathInfo
}

export interface ISwaggerPathInfo {
  description?: string
  summary: string
  operationId: string
  parameters?: Array<ISwaggerParameterDefaults & ISwaggerSchemeObject>
  tags: string[]
  responses: ISwaggerPathResponses
}

export interface ISwaggerParameterDefaults {
  name: string
  in: SwaggerParameterIn
  schema?: ISwaggerScheme
}

export interface ISwaggerPathResponses {
  [statusCode: string]: ISwaggerPathResponse
}

export interface ISwaggerPathResponse {
  schema: ISwaggerScheme
  description: string
}

export interface ISwaggerSchemeObjectProperty {
  [key: string]: ISwaggerScheme
}

export interface ISwaggerSchemeObjectProperties {
  properties: ISwaggerSchemeObjectProperty
  required: string[]
  type: SwaggerSchemeType.object
}

export interface ISwaggerSchemeObject extends Partial<ISwaggerSchemeObjectProperties> {
  type: SwaggerSchemeType.object
  allOf?: ISwaggerSchemeObjectProperties[]
}

export interface ISwaggerSchemeBoolean {
  type: SwaggerSchemeType.boolean
}

export interface ISwaggerSchemeArray {
  type: SwaggerSchemeType.array
  items: ISwaggerScheme
}

export interface ISwaggerSchemeString {
  type: SwaggerSchemeType.string
  enum?: string[]
}

export interface ISwaggerSchemeNumber {
  type: SwaggerSchemeType.number | SwaggerSchemeType.integer
}

export type ISwaggerScheme =
  | ISwaggerSchemeObject
  | ISwaggerSchemeBoolean
  | ISwaggerSchemeArray
  | ISwaggerSchemeString
  | ISwaggerSchemeNumber
