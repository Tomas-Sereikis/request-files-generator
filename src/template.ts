import * as assert from 'assert'
import { JoiCodeGenerator } from './generators/JoiCodeGenerator'
import {
  ISwaggerPathInfo,
  ISwaggerPaths,
  ISwaggerSchemeObject,
  ISwaggerSchemeObjectProperty,
  SwaggerParameterIn,
  SwaggerSchemeType,
} from './swagger'
import { TypeScriptCodeGenerator } from './generators/TypeScriptCodeGenerator'
import { camelCase, upperFirst, template } from 'lodash'
import { formatCode } from './prettier'

export interface ICreateTemplateParams {
  paths: ISwaggerPaths
  url: string
  method: string
  statusCode: string
  template: string
}

const removeNewLines = (str: string) => str.replace(/\r?\n|\r/g, ' ')

function parametersInToObjectScheme(
  parameters: ISwaggerPathInfo['parameters'],
  inType: SwaggerParameterIn,
): { count: number; required: string[]; scheme: ISwaggerSchemeObject } {
  const properties = parameters
    .filter(curr => curr.in === inType)
    .reduce(
      (previousValue, currentValue) => {
        if (
          currentValue.in === SwaggerParameterIn.body && 
          currentValue.name === 'body' && 
          currentValue.schema && 
          currentValue.schema.type === SwaggerSchemeType.object
        ) {
          return {
            ...previousValue,
            ...currentValue.schema.properties,
          }
        } else {
          return {
            ...previousValue,
            [currentValue.name]: currentValue.schema || currentValue,
          }
        }
      },
      {} as ISwaggerSchemeObjectProperty,
    )
  const required = parameters.filter(curr => curr.in === inType && curr.required).map(curr => curr.name)
  return {
    count: Object.keys(properties).length,
    required,
    scheme: {
      type: SwaggerSchemeType.object,
      properties,
      required,
    },
  }
}

function applyUrlParamsBuilder(url: string) {
  return (variable: string) => {
    const replaced = url.replace(/{(.*?)}/g, (match, value) => `\${${variable}.path.${value}}`)
    return `\`${replaced}\``
  }
}

export function createTemplate(options: ICreateTemplateParams): string {
  const tsCodeGenerator = new TypeScriptCodeGenerator()
  const joiCodeGenerator = new JoiCodeGenerator()

  const endpoint = options.paths[options.url][options.method]
  const response = endpoint ? endpoint.responses[options.statusCode] : null

  assert(!!endpoint, 'Swagger path info is not found!')
  assert(!!response, 'Swagger response is not found!')

  const bodyObjectScheme = parametersInToObjectScheme(endpoint.parameters, SwaggerParameterIn.body)
  const queryObjectScheme = parametersInToObjectScheme(endpoint.parameters, SwaggerParameterIn.query)
  const paramsObjectScheme = parametersInToObjectScheme(endpoint.parameters, SwaggerParameterIn.path)

  let optionsScheme = {}
  let required = []
  if (bodyObjectScheme.count) {
    optionsScheme = { ...optionsScheme, body: bodyObjectScheme.scheme }
    required.push(SwaggerParameterIn.body)
  }
  if (queryObjectScheme.count) {
    optionsScheme = { ...optionsScheme, query: queryObjectScheme.scheme }
  }
  if (queryObjectScheme.count && queryObjectScheme.required.length) {
    required.push(SwaggerParameterIn.query)
  }
  if (paramsObjectScheme.count) {
    optionsScheme = { ...optionsScheme, path: paramsObjectScheme.scheme }
    required.push(SwaggerParameterIn.path)
  }

  const optionsObject = tsCodeGenerator.onInterface(optionsScheme, required)
  const templateBuilder = template(options.template)
  const operationId = camelCase(endpoint.operationId)
  const description = endpoint.description ? removeNewLines(endpoint.description) : ''

  return formatCode(
    templateBuilder({
      applyUrlParams: applyUrlParamsBuilder(options.url),
      description: description,
      containsBody: ['post', 'put', 'patch'].includes(options.method),
      containsQuery: !!queryObjectScheme.count,
      responseInterface: tsCodeGenerator.of(response!.schema),
      optionsInterface: optionsObject,
      scheme: joiCodeGenerator.of(response!.schema),
      isDelete: ['delete'].includes(options.method),
      operationId,
      method: options.method,
      upperOperationId: upperFirst(operationId),
    }),
  )
}
