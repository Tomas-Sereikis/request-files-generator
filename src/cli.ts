import * as assert from 'assert'
import * as prompts from 'prompts'
import * as path from 'path'
import * as swagger from './swagger'
import { promises as fs } from 'fs'
import { createTemplate } from './template'

const MSG_OVERWRITE_FILE = 'Would you like to overwrite {fileName}'
const MSG_ENTER_TEMPLATE_PATH = 'Enter template path'
const MSG_ENTER_OUTPUT_PATH = 'Enter output path'
const MSG_ENTER_SWAGGER_JSON_URL = 'Enter swagger.json file url'
const MSG_SELECT_PATHS = 'Select paths that you would like to use'
const MSG_SELECT_OPERATIONS = 'Select operations that you would like to use'
const MSG_SELECT_STATUS_CODE = 'Select statusCode for {operationId}'

main(process.argv.slice(2)).catch(console.error)

async function main(argv: ReadonlyArray<string>) {
  const { value: templatePath } = await prompts({
    message: MSG_ENTER_TEMPLATE_PATH,
    name: 'value',
    type: 'text',
  })

  const templateStat = await fs.stat(templatePath)
  assert(templateStat.isFile(), 'Template file dose not exist!')
  const templateContent = String(await fs.readFile(templatePath))

  const { value: outputPath } = await prompts({
    message: MSG_ENTER_OUTPUT_PATH,
    name: 'value',
    type: 'text',
  })

  const outputStat = await fs.stat(outputPath)
  assert(outputStat.isDirectory(), 'Output directory dose not exist!')

  const { value: swaggerUrl } = await prompts({
    message: MSG_ENTER_SWAGGER_JSON_URL,
    name: 'value',
    type: 'text',
  })

  const docs = await swagger.validate(swaggerUrl)
  const choicesOfPaths = Object.keys(docs.paths).map(path => ({ value: path, title: path }))
  const { value: paths } = await prompts({
    message: MSG_SELECT_PATHS,
    name: 'value',
    type: 'multiselect',
    choices: choicesOfPaths,
  })

  const choicesOfOperations = Object.keys(docs.paths)
    .filter(path => paths.includes(path))
    .map(path => Object.values(docs.paths[path]))
    .reduce(
      (previousValue, currentValue) => [...previousValue, ...currentValue.map(curr => curr.operationId)],
      [] as string[],
    )
    .map(operation => ({ value: operation, title: operation }))

  const { value: operations } = await prompts({
    message: MSG_SELECT_OPERATIONS,
    name: 'value',
    type: 'multiselect',
    choices: choicesOfOperations,
  })

  for (const url of Object.keys(docs.paths)) {
    for (const method of Object.keys(docs.paths[url])) {
      const info = docs.paths[url][method]
      if (!operations.includes(info.operationId)) {
        continue
      }
      const statusCodes = Object.keys(info.responses)
      let statusCode: string | undefined
      if (statusCodes.length === 0) {
        continue
      } else if (statusCodes.length === 1) {
        statusCode = statusCodes[0]
      } else {
        const choicesOfStatusCodes = statusCodes.map(code => ({ value: code, title: code }))
        const selection = await prompts({
          message: MSG_SELECT_STATUS_CODE.replace('{operationId}', info.operationId),
          name: 'value',
          type: 'select',
          choices: choicesOfStatusCodes,
        })
        statusCode = selection.value
      }
      const template = await createTemplate({
        template: templateContent,
        paths: docs.paths,
        url,
        method,
        statusCode: statusCode!,
      })
      const filePath = path.resolve(outputPath, `${info.operationId}.ts`)
      const filePathStat = await fs.stat(filePath).catch(() => null)
      if (filePathStat && filePathStat.isFile()) {
        const { value: overwrite } = await prompts({
          message: MSG_OVERWRITE_FILE.replace('{filePath}', filePath),
          name: 'value',
          type: 'toggle',
          initial: true,
          active: 'yes',
          inactive: 'no',
        })
        if (overwrite === 'no') {
          continue
        }
      }
      await fs.writeFile(filePath, template)
    }
  }
}
