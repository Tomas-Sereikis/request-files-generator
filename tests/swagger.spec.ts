import * as swagger from '../src/swagger'
import mockOfCategories from './mocks/categories'
import mockOfChannels from './mocks/channel'
import mockOfPetstore from './mocks/petstore'

test('should test categories parsing paths', async () => {
  const data = await swagger.validate(mockOfCategories)
  expect(data).toMatchSnapshot()
})

test('should test channels parsing paths', async () => {
  const data = await swagger.validate(mockOfChannels)
  expect(data).toMatchSnapshot()
})

test('should test petstore parsing paths', async () => {
  const data = await swagger.validate(mockOfPetstore)
  expect(data).toMatchSnapshot()
})
