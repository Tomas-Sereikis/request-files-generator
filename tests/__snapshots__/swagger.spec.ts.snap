// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should test categories parsing paths 1`] = `
Object {
  "definitions": Object {
    "ChannelCategoriesFetchAllResponse": Object {
      "properties": Object {
        "data": Object {
          "items": Object {
            "properties": Object {
              "color": Object {
                "type": "string",
              },
              "id": Object {
                "type": "string",
              },
              "region": Object {
                "type": "string",
              },
              "sortIndex": Object {
                "type": "number",
              },
              "svg": Object {
                "type": "string",
              },
              "title": Object {
                "type": "string",
              },
            },
            "required": Array [
              "id",
              "region",
              "sortIndex",
              "svg",
              "title",
            ],
            "type": "object",
          },
          "type": "array",
        },
        "next": Object {
          "type": "boolean",
        },
      },
      "required": Array [
        "data",
      ],
      "type": "object",
    },
    "ChannelCategory": Object {
      "properties": Object {
        "color": Object {
          "type": "string",
        },
        "id": Object {
          "type": "string",
        },
        "region": Object {
          "type": "string",
        },
        "sortIndex": Object {
          "type": "number",
        },
        "svg": Object {
          "type": "string",
        },
        "title": Object {
          "type": "string",
        },
      },
      "required": Array [
        "id",
        "region",
        "sortIndex",
        "svg",
        "title",
      ],
      "type": "object",
    },
  },
  "info": Object {
    "title": "content",
    "version": "1.0.0",
  },
  "paths": Object {
    "/channel-categories": Object {
      "get": Object {
        "operationId": "getChannelCategories",
        "parameters": Array [
          Object {
            "description": "Region",
            "in": "query",
            "name": "region",
            "type": "string",
          },
        ],
        "responses": Object {
          "200": Object {
            "description": "Successful",
            "schema": Object {
              "properties": Object {
                "data": Object {
                  "items": Object {
                    "properties": Object {
                      "color": Object {
                        "type": "string",
                      },
                      "id": Object {
                        "type": "string",
                      },
                      "region": Object {
                        "type": "string",
                      },
                      "sortIndex": Object {
                        "type": "number",
                      },
                      "svg": Object {
                        "type": "string",
                      },
                      "title": Object {
                        "type": "string",
                      },
                    },
                    "required": Array [
                      "id",
                      "region",
                      "sortIndex",
                      "svg",
                      "title",
                    ],
                    "type": "object",
                  },
                  "type": "array",
                },
                "next": Object {
                  "type": "boolean",
                },
              },
              "required": Array [
                "data",
              ],
              "type": "object",
            },
          },
        },
        "summary": "Fetch channel categories list",
        "tags": Array [
          "channel-categories",
        ],
      },
    },
  },
  "swagger": "2.0",
}
`;

exports[`should test channels parsing paths 1`] = `
Object {
  "basePath": "/",
  "definitions": Object {
    "Channel": Object {
      "properties": Object {
        "avatarImageId": Object {
          "type": "string",
        },
        "coverImageId": Object {
          "type": "string",
        },
        "creatorId": Object {
          "type": "string",
        },
        "datetime": Object {
          "format": "date",
          "type": "string",
        },
        "description": Object {
          "type": "string",
        },
        "followers": Object {
          "type": "number",
        },
        "id": Object {
          "type": "string",
        },
        "playlistMode": Object {
          "enum": Array [
            "new",
            "shuffle",
          ],
          "type": "string",
        },
        "private": Object {
          "type": "boolean",
        },
        "showInNews": Object {
          "type": "boolean",
        },
        "title": Object {
          "type": "string",
        },
        "website": Object {
          "type": "string",
        },
      },
      "required": Array [
        "avatarImageId",
        "coverImageId",
        "creatorId",
        "datetime",
        "description",
        "followers",
        "id",
        "playlistMode",
        "private",
        "showInNews",
        "title",
      ],
      "type": "object",
    },
    "ChannelsCreateRequest": Object {
      "properties": Object {
        "avatarImageId": Object {
          "type": "string",
        },
        "coverImageId": Object {
          "type": "string",
        },
        "creatorId": Object {
          "type": "string",
        },
        "description": Object {
          "type": "string",
        },
        "playlistMode": Object {
          "default": "shuffle",
          "enum": Array [
            "new",
            "shuffle",
          ],
          "type": "string",
        },
        "private": Object {
          "type": "boolean",
        },
        "title": Object {
          "type": "string",
        },
        "website": Object {
          "type": "string",
        },
      },
      "required": Array [
        "avatarImageId",
        "coverImageId",
        "creatorId",
        "description",
        "private",
        "title",
      ],
      "type": "object",
    },
    "ChannelsFetchAllList": Object {
      "items": Object {
        "properties": Object {
          "avatarImageId": Object {
            "type": "string",
          },
          "coverImageId": Object {
            "type": "string",
          },
          "creatorId": Object {
            "type": "string",
          },
          "datetime": Object {
            "format": "date",
            "type": "string",
          },
          "description": Object {
            "type": "string",
          },
          "followers": Object {
            "type": "number",
          },
          "id": Object {
            "type": "string",
          },
          "playlistMode": Object {
            "enum": Array [
              "new",
              "shuffle",
            ],
            "type": "string",
          },
          "private": Object {
            "type": "boolean",
          },
          "showInNews": Object {
            "type": "boolean",
          },
          "title": Object {
            "type": "string",
          },
          "website": Object {
            "type": "string",
          },
        },
        "required": Array [
          "avatarImageId",
          "coverImageId",
          "creatorId",
          "datetime",
          "description",
          "followers",
          "id",
          "playlistMode",
          "private",
          "showInNews",
          "title",
        ],
        "type": "object",
      },
      "type": "array",
    },
    "ChannelsFetchAllResponse": Object {
      "properties": Object {
        "data": Object {
          "items": Object {
            "properties": Object {
              "avatarImageId": Object {
                "type": "string",
              },
              "coverImageId": Object {
                "type": "string",
              },
              "creatorId": Object {
                "type": "string",
              },
              "datetime": Object {
                "format": "date",
                "type": "string",
              },
              "description": Object {
                "type": "string",
              },
              "followers": Object {
                "type": "number",
              },
              "id": Object {
                "type": "string",
              },
              "playlistMode": Object {
                "enum": Array [
                  "new",
                  "shuffle",
                ],
                "type": "string",
              },
              "private": Object {
                "type": "boolean",
              },
              "showInNews": Object {
                "type": "boolean",
              },
              "title": Object {
                "type": "string",
              },
              "website": Object {
                "type": "string",
              },
            },
            "required": Array [
              "avatarImageId",
              "coverImageId",
              "creatorId",
              "datetime",
              "description",
              "followers",
              "id",
              "playlistMode",
              "private",
              "showInNews",
              "title",
            ],
            "type": "object",
          },
          "type": "array",
        },
        "next": Object {
          "type": "boolean",
        },
      },
      "required": Array [
        "data",
      ],
      "type": "object",
    },
    "ChannelsFetchByIdResponse": Object {
      "properties": Object {
        "data": Object {
          "properties": Object {
            "avatarImageId": Object {
              "type": "string",
            },
            "coverImageId": Object {
              "type": "string",
            },
            "creatorId": Object {
              "type": "string",
            },
            "datetime": Object {
              "format": "date",
              "type": "string",
            },
            "description": Object {
              "type": "string",
            },
            "followers": Object {
              "type": "number",
            },
            "id": Object {
              "type": "string",
            },
            "playlistMode": Object {
              "enum": Array [
                "new",
                "shuffle",
              ],
              "type": "string",
            },
            "private": Object {
              "type": "boolean",
            },
            "showInNews": Object {
              "type": "boolean",
            },
            "title": Object {
              "type": "string",
            },
            "website": Object {
              "type": "string",
            },
          },
          "required": Array [
            "avatarImageId",
            "coverImageId",
            "creatorId",
            "datetime",
            "description",
            "followers",
            "id",
            "playlistMode",
            "private",
            "showInNews",
            "title",
          ],
          "type": "object",
        },
        "next": Object {
          "type": "boolean",
        },
      },
      "required": Array [
        "data",
      ],
      "type": "object",
    },
    "Wakie": Object {
      "properties": Object {
        "audioId": Object {
          "type": "string",
        },
        "authorName": Object {
          "type": "string",
        },
        "channelId": Object {
          "type": "string",
        },
        "comments": Object {
          "type": "number",
        },
        "creatorId": Object {
          "type": "string",
        },
        "datetime": Object {
          "format": "date",
          "type": "string",
        },
        "duration": Object {
          "type": "number",
        },
        "id": Object {
          "type": "string",
        },
        "imageId": Object {
          "type": "string",
        },
        "likes": Object {
          "type": "number",
        },
        "plays": Object {
          "type": "number",
        },
        "publishDatetime": Object {
          "format": "date",
          "type": "string",
        },
        "status": Object {
          "enum": Array [
            "created",
            "finished",
            "error",
          ],
          "type": "string",
        },
        "title": Object {
          "type": "string",
        },
      },
      "required": Array [
        "audioId",
        "authorName",
        "channelId",
        "comments",
        "creatorId",
        "datetime",
        "duration",
        "id",
        "imageId",
        "likes",
        "plays",
        "publishDatetime",
        "status",
        "title",
      ],
      "type": "object",
    },
  },
  "info": Object {
    "title": "content",
    "version": "1.0.0",
  },
  "paths": Object {
    "/channels": Object {
      "get": Object {
        "operationId": "getChannels",
        "parameters": Array [
          Object {
            "description": "Category id",
            "in": "query",
            "name": "categoryId",
            "type": "string",
          },
          Object {
            "description": "Creator id",
            "in": "query",
            "name": "creatorId",
            "type": "string",
          },
          Object {
            "collectionFormat": "multi",
            "description": "List of id's",
            "in": "query",
            "items": Object {
              "type": "string",
            },
            "name": "id",
            "type": "array",
          },
          Object {
            "default": false,
            "description": "Private channels included",
            "in": "query",
            "name": "private",
            "type": "boolean",
          },
          Object {
            "description": "Region name",
            "in": "query",
            "name": "region",
            "type": "string",
          },
        ],
        "responses": Object {
          "200": Object {
            "description": "Successful",
            "schema": Object {
              "properties": Object {
                "data": Object {
                  "items": Object {
                    "properties": Object {
                      "avatarImageId": Object {
                        "type": "string",
                      },
                      "coverImageId": Object {
                        "type": "string",
                      },
                      "creatorId": Object {
                        "type": "string",
                      },
                      "datetime": Object {
                        "format": "date",
                        "type": "string",
                      },
                      "description": Object {
                        "type": "string",
                      },
                      "followers": Object {
                        "type": "number",
                      },
                      "id": Object {
                        "type": "string",
                      },
                      "playlistMode": Object {
                        "enum": Array [
                          "new",
                          "shuffle",
                        ],
                        "type": "string",
                      },
                      "private": Object {
                        "type": "boolean",
                      },
                      "showInNews": Object {
                        "type": "boolean",
                      },
                      "title": Object {
                        "type": "string",
                      },
                      "website": Object {
                        "type": "string",
                      },
                    },
                    "required": Array [
                      "avatarImageId",
                      "coverImageId",
                      "creatorId",
                      "datetime",
                      "description",
                      "followers",
                      "id",
                      "playlistMode",
                      "private",
                      "showInNews",
                      "title",
                    ],
                    "type": "object",
                  },
                  "type": "array",
                },
                "next": Object {
                  "type": "boolean",
                },
              },
              "required": Array [
                "data",
              ],
              "type": "object",
            },
          },
        },
        "summary": "Fetch channels list",
        "tags": Array [
          "channels",
        ],
      },
      "post": Object {
        "operationId": "postChannels",
        "parameters": Array [
          Object {
            "in": "body",
            "name": "body",
            "schema": Object {
              "properties": Object {
                "avatarImageId": Object {
                  "type": "string",
                },
                "coverImageId": Object {
                  "type": "string",
                },
                "creatorId": Object {
                  "type": "string",
                },
                "description": Object {
                  "type": "string",
                },
                "playlistMode": Object {
                  "default": "shuffle",
                  "enum": Array [
                    "new",
                    "shuffle",
                  ],
                  "type": "string",
                },
                "private": Object {
                  "type": "boolean",
                },
                "title": Object {
                  "type": "string",
                },
                "website": Object {
                  "type": "string",
                },
              },
              "required": Array [
                "avatarImageId",
                "coverImageId",
                "creatorId",
                "description",
                "private",
                "title",
              ],
              "type": "object",
            },
          },
        ],
        "responses": Object {
          "200": Object {
            "description": "Successful",
            "schema": Object {
              "properties": Object {
                "data": Object {
                  "properties": Object {
                    "avatarImageId": Object {
                      "type": "string",
                    },
                    "coverImageId": Object {
                      "type": "string",
                    },
                    "creatorId": Object {
                      "type": "string",
                    },
                    "datetime": Object {
                      "format": "date",
                      "type": "string",
                    },
                    "description": Object {
                      "type": "string",
                    },
                    "followers": Object {
                      "type": "number",
                    },
                    "id": Object {
                      "type": "string",
                    },
                    "playlistMode": Object {
                      "enum": Array [
                        "new",
                        "shuffle",
                      ],
                      "type": "string",
                    },
                    "private": Object {
                      "type": "boolean",
                    },
                    "showInNews": Object {
                      "type": "boolean",
                    },
                    "title": Object {
                      "type": "string",
                    },
                    "website": Object {
                      "type": "string",
                    },
                  },
                  "required": Array [
                    "avatarImageId",
                    "coverImageId",
                    "creatorId",
                    "datetime",
                    "description",
                    "followers",
                    "id",
                    "playlistMode",
                    "private",
                    "showInNews",
                    "title",
                  ],
                  "type": "object",
                },
                "next": Object {
                  "type": "boolean",
                },
              },
              "required": Array [
                "data",
              ],
              "type": "object",
            },
          },
        },
        "summary": "Create channel",
        "tags": Array [
          "channels",
        ],
      },
    },
  },
  "schemes": Array [
    "https",
  ],
  "swagger": "2.0",
  "tags": Array [],
}
`;

exports[`should test petstore parsing paths 1`] = `
Object {
  "basePath": "/api",
  "definitions": Object {
    "Error": Object {
      "properties": Object {
        "code": Object {
          "format": "int32",
          "type": "integer",
        },
        "message": Object {
          "type": "string",
        },
      },
      "required": Array [
        "code",
        "message",
      ],
      "type": "object",
    },
    "NewPet": Object {
      "properties": Object {
        "name": Object {
          "type": "string",
        },
        "tag": Object {
          "type": "string",
        },
      },
      "required": Array [
        "name",
      ],
      "type": "object",
    },
    "Pet": Object {
      "allOf": Array [
        Object {
          "properties": Object {
            "name": Object {
              "type": "string",
            },
            "tag": Object {
              "type": "string",
            },
          },
          "required": Array [
            "name",
          ],
          "type": "object",
        },
        Object {
          "properties": Object {
            "id": Object {
              "format": "int64",
              "type": "integer",
            },
          },
          "required": Array [
            "id",
          ],
        },
      ],
      "type": "object",
    },
  },
  "host": "petstore.swagger.io",
  "info": Object {
    "description": "A sample API that uses a petstore as an example to demonstrate features in the swagger-2.0 specification",
    "title": "Swagger Petstore",
    "version": "1.0.0",
  },
  "paths": Object {
    "/pets": Object {
      "get": Object {
        "description": "Returns all pets from the system that the user has access to Nam sed condimentum est.",
        "operationId": "findPets",
        "parameters": Array [
          Object {
            "collectionFormat": "csv",
            "description": "tags to filter by",
            "in": "query",
            "items": Object {
              "type": "string",
            },
            "name": "tags",
            "required": false,
            "type": "array",
          },
          Object {
            "description": "maximum number of results to return",
            "format": "int32",
            "in": "query",
            "name": "limit",
            "required": false,
            "type": "integer",
          },
        ],
        "responses": Object {
          "200": Object {
            "description": "pet response",
            "schema": Object {
              "items": Object {
                "allOf": Array [
                  Object {
                    "properties": Object {
                      "name": Object {
                        "type": "string",
                      },
                      "tag": Object {
                        "type": "string",
                      },
                    },
                    "required": Array [
                      "name",
                    ],
                    "type": "object",
                  },
                  Object {
                    "properties": Object {
                      "id": Object {
                        "format": "int64",
                        "type": "integer",
                      },
                    },
                    "required": Array [
                      "id",
                    ],
                  },
                ],
                "type": "object",
              },
              "type": "array",
            },
          },
          "default": Object {
            "description": "unexpected error",
            "schema": Object {
              "properties": Object {
                "code": Object {
                  "format": "int32",
                  "type": "integer",
                },
                "message": Object {
                  "type": "string",
                },
              },
              "required": Array [
                "code",
                "message",
              ],
              "type": "object",
            },
          },
        },
      },
      "post": Object {
        "description": "Creates a new pet in the store.  Duplicates are allowed",
        "operationId": "addPet",
        "parameters": Array [
          Object {
            "description": "Pet to add to the store",
            "in": "body",
            "name": "pet",
            "required": true,
            "schema": Object {
              "properties": Object {
                "name": Object {
                  "type": "string",
                },
                "tag": Object {
                  "type": "string",
                },
              },
              "required": Array [
                "name",
              ],
              "type": "object",
            },
          },
        ],
        "responses": Object {
          "200": Object {
            "description": "pet response",
            "schema": Object {
              "allOf": Array [
                Object {
                  "properties": Object {
                    "name": Object {
                      "type": "string",
                    },
                    "tag": Object {
                      "type": "string",
                    },
                  },
                  "required": Array [
                    "name",
                  ],
                  "type": "object",
                },
                Object {
                  "properties": Object {
                    "id": Object {
                      "format": "int64",
                      "type": "integer",
                    },
                  },
                  "required": Array [
                    "id",
                  ],
                },
              ],
              "type": "object",
            },
          },
          "default": Object {
            "description": "unexpected error",
            "schema": Object {
              "properties": Object {
                "code": Object {
                  "format": "int32",
                  "type": "integer",
                },
                "message": Object {
                  "type": "string",
                },
              },
              "required": Array [
                "code",
                "message",
              ],
              "type": "object",
            },
          },
        },
      },
    },
    "/pets/{id}": Object {
      "delete": Object {
        "description": "deletes a single pet based on the ID supplied",
        "operationId": "deletePet",
        "parameters": Array [
          Object {
            "description": "ID of pet to delete",
            "format": "int64",
            "in": "path",
            "name": "id",
            "required": true,
            "type": "integer",
          },
        ],
        "responses": Object {
          "204": Object {
            "description": "pet deleted",
          },
          "default": Object {
            "description": "unexpected error",
            "schema": Object {
              "properties": Object {
                "code": Object {
                  "format": "int32",
                  "type": "integer",
                },
                "message": Object {
                  "type": "string",
                },
              },
              "required": Array [
                "code",
                "message",
              ],
              "type": "object",
            },
          },
        },
      },
      "get": Object {
        "description": "Returns a user based on a single ID, if the user does not have access to the pet",
        "operationId": "find pet by id",
        "parameters": Array [
          Object {
            "description": "ID of pet to fetch",
            "format": "int64",
            "in": "path",
            "name": "id",
            "required": true,
            "type": "integer",
          },
        ],
        "responses": Object {
          "200": Object {
            "description": "pet response",
            "schema": Object {
              "allOf": Array [
                Object {
                  "properties": Object {
                    "name": Object {
                      "type": "string",
                    },
                    "tag": Object {
                      "type": "string",
                    },
                  },
                  "required": Array [
                    "name",
                  ],
                  "type": "object",
                },
                Object {
                  "properties": Object {
                    "id": Object {
                      "format": "int64",
                      "type": "integer",
                    },
                  },
                  "required": Array [
                    "id",
                  ],
                },
              ],
              "type": "object",
            },
          },
          "default": Object {
            "description": "unexpected error",
            "schema": Object {
              "properties": Object {
                "code": Object {
                  "format": "int32",
                  "type": "integer",
                },
                "message": Object {
                  "type": "string",
                },
              },
              "required": Array [
                "code",
                "message",
              ],
              "type": "object",
            },
          },
        },
      },
    },
  },
  "swagger": "2.0",
}
`;
